posts/2017-12-07-parser-combinators.md

```diff
@@ -0,0 +1,286 @@
+
+(value prog-name-p "gyxo") ; :gyxo
+```
+
+After the program name comes the value in parenthesis. Luckily we have a handy function to parse between parenthesis.
+
+
+``` clojure 
+(value (parens dec-num) "(22)") ; 22
+```
+
+We need both values, the program name and the value in parenthesis:
+
+``` clojure
+(def prog-val-p (<*> prog-name-p (skip-ws (parens dec-num))))
+
+(value prog-val-p "abcd (67)") ; [:abcd 67]
+```
+
+First case done, now we need to parse a program that has "dependencies".
+
+Something like `abcd (88) -> ktlj, cntj, xhth`.
+
+As we did before, let us split the task. We do have a parser for a program with value, so let us focus on a parser for the dependencies.
+
+Here the dependencies are a list of comma separated _program name_. Once again, we can use one of the already defined parser combinators `sep-by`.
+
+``` clojure
+(def dep-list-p (sep-by comma prog-name-p))
+
+(value dep-list-p "abcd, efgh, ijkl, mnop") ; [:abcd :efgh :ijkl :mnop]
+```
+
+To combine both we need a ` -> ` in the middle. We could use the primitive parser `token` that does exactly that.
+
+Also, because we only care about the list and not the token, we can use the `>>` combinator to get the value of the last parser only.
+
+``` clojure
+(def assign-p (>> (skip-ws (token "->")) (skip-ws dep-list-p)))
+
+(value assign-p " -> aaaa, bbbb, cccc") ; [:aaaa :bbbb :cccc]
+```
+
+Almost ready! Now we need to combine two scenarios. One where we have only the program name and value and the other where also there are dependencies.
+
+To model that we could use the two parsers we already have, making the dependencies one _optional_.
+
+``` clojure
+(def line-p (<*> prog-val-p (optional assign-p)))
+
+(value line-p "abcd (87)") ; [[:abcd 87] nil]
+
+(value line-p "abcd (87) -> aaaa, bbbb, cccc") ; [[:abcd 87] [:aaaa :bbbb :cccc]]
+
+```
+
+Another way to define the combination of both parsers is to use `bind` in the form of `let` so each _name_ is bound to the result of each parser.
+
+Instead of having the concatenations of two lists (and maybe `nil` as a second element) we could concatenate all the values together as part of the same list.
+
+``` clojure
+(def line-p
+  (bind [prg prog-val-p
+         deps (optional assign-p)]
+         (return (concat prg (or deps [])))))
+
+(value line-p "abcd (87)") ; (:abcd 87)
+
+(value line-p "abcd (87) -> aaaa, bbbb, cccc") ; (:abcd 87 :aaaa :bbbb :cccc)
+
+```
+
+## Is it worth it?
+
+Here is a summary of the solution:
+
+``` clojure
+(def prog-name-p (<$> #(->> % (apply str) keyword) (times 4 letter)))
+
+(def prog-val-p (<*> prog-name-p (skip-ws (parens dec-num))))
+
+(def dep-list-p (sep-by comma prog-name-p))
+
+(def assign-p (>> (skip-ws (token "->")) (skip-ws dep-list-p)))
+
+(def line-p
+  (bind [prg prog-val-p
+         deps (optional assign-p)]
+         (return (concat prg (or deps [])))))
+
+```
+
+Clearly is much more code than a regular expression but it has many benefits over a reg exp.
```

Comment:
Clearly `this` is much more code than a regular expression but it has many benefits over a reg exp.

Reply:
fixed!